File: built-in plugins dsp/mb-drums-rock.cpp
/**
 * MB Rock Kit
 * Category : instrument
 * Type     : drums
 * Version  : 1.0.0
 * Author   : Max Booster
 * Desc     : Powerful rock drum kit
 *
 * Part of Max Booster Built-In Plugins DSP
 */

#ifndef MB_DRUMS_ROCK_H
#define MB_DRUMS_ROCK_H

#include <algorithm>
#include <cmath>
#include <cstring>
#include "PluginBase.h"

class MbDrumsRock : public PluginBase {
public:
    static constexpr const char* PLUGIN_ID      = "mb-drums-rock";
    static constexpr const char* PLUGIN_NAME    = "MB Rock Kit";
    static constexpr const char* PLUGIN_TYPE    = "drums";
    static constexpr const char* PLUGIN_CATEGORY = "instrument";
    static constexpr const char* VERSION         = "1.0.0";

    struct Parameters {
    float attack = 0.9f;  // range [0, 1]
    float volume = 0.8f;  // range [0, 1]
    };

    MbDrumsRock() = default;
    ~MbDrumsRock() override = default;

    void setSampleRate(double sampleRate) override {
        sampleRate_ = sampleRate;
        reset();
    }

    void reset() override {
        std::memset(buffer_, 0, sizeof(buffer_));
    }

    void process(float* left, float* right, int numSamples, Parameters params) {
        params.attack = std::clamp(params.attack, 0f, 1f);
        params.volume = std::clamp(params.volume, 0f, 1f);
        for (int i = 0; i < numSamples; ++i) {
            left[i]  = processSample(left[i],  params);
            right[i] = processSample(right[i], params);
        }
    }

private:
    double sampleRate_ = 44100.0;
    float  buffer_[65536] = {};

    inline float processSample(float input, const Parameters& params) {
        // DSP implementation for MB Rock Kit
        return input;
    }
};

#endif // MB_DRUMS_ROCK_H

File: built-in plugins dsp/mb-drums-rock.ts
import type { PluginDefinition } from "../server/services/pluginHostService";

const MbDrumsRockPlugin: PluginDefinition = {
  id: "mb-drums-rock",
  slug: "mb-drums-rock",
  name: "MB Rock Kit",
  category: "instrument",
  type: "drums",
  version: "1.0.0",
  description: "Powerful rock drum kit",
  author: "Max Booster",
  grade: "A",
  oscillators: [
    { type: "sine", detune: 0, gain: 0.9 },
    { type: "noise", detune: 0, gain: 0.6 },
  ],
  envelope: { attack: 0.001, decay: 0.15, sustain: 0.1, release: 0.25 },
  parameters: [
    {
      id: "attack",
      name: "Attack",
      type: "float",
      defaultValue: 0.9,
      minValue: 0,
      maxValue: 1,
      automatable: true,
    },
    {
      id: "volume",
      name: "Volume",
      type: "float",
      defaultValue: 0.8,
      minValue: 0,
      maxValue: 1,
      automatable: true,
    },
  ],
  defaultPreset: { attack: 0.9, volume: 0.8 },
};

export default MbDrumsRockPlugin;
